File: reactor/java-reactor/src/main/java/com/example/reactor/javareactor/handler/UserHandler.java
package com.example.reactor.javareactor.handler;

import com.example.reactor.javareactor.dto.request.CreateUserRequest;
import com.example.reactor.javareactor.dto.response.UserResponse;
import com.example.reactor.javareactor.service.UserService;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Component;
import org.springframework.web.reactive.function.server.ServerRequest;
import org.springframework.web.reactive.function.server.ServerResponse;
import reactor.core.publisher.Mono;

@RequiredArgsConstructor
@Component
public class UserHandler {
    private final UserService userService;

    public Mono<ServerResponse> getUserById(ServerRequest request) {
        return ServerResponse.ok()
                             .body(userService.getUserById(request.pathVariable("id")), UserResponse.class);
    }

    public Mono<ServerResponse> createUser(ServerRequest request) {
        return request.bodyToMono(CreateUserRequest.class)
                      .flatMap(createUserRequest ->
                                   ServerResponse.ok()
                                                 .body(userService.createUser(createUserRequest), UserResponse.class)
                      );
    }
}

File: reactor/kotlin-reactor/src/main/kotlin/com/example/reactor/kotlinreactor/handler/UserHandler.kt
package com.example.reactor.kotlinreactor.handler

import com.example.reactor.kotlinreactor.dto.request.CreateUserRequest
import com.example.reactor.kotlinreactor.service.UserService
import org.springframework.stereotype.Component
import org.springframework.web.reactive.function.server.ServerRequest
import org.springframework.web.reactive.function.server.ServerResponse
import org.springframework.web.reactive.function.server.body
import org.springframework.web.reactive.function.server.bodyToMono
import reactor.core.publisher.Mono

@Component
class UserHandler(
    private val userService: UserService
) {
    fun getUserById(request: ServerRequest): Mono<ServerResponse> =
        ServerResponse.ok()
            .body(userService.getUserById(request.pathVariable("id")))

    fun createUser(request: ServerRequest): Mono<ServerResponse> =
        request.bodyToMono<CreateUserRequest>()
            .flatMap {
                ServerResponse.ok()
                    .body(userService.createUser(it))
            }
}
